README.md

```diff
@@ -1,19 +1,122 @@
+
+This is an early prototype, both in terms of the engine performance and the API and toolbox for the library user. The user should be ready to add missing primitives, as well as obvious tools that should be predefined but aren't. One can already build with the library all basic neural network architectures, such as fully connected, recurrent, convolutional and residual networks. The library, due to its loose coupling of differentiation and data containers, can naturally handle exotic variants of such networks that wouldn't express well in a language of matrices or tensors. 
+
+Applications outside machine learning are very much in scope, given that the notion of a neural network is not hardwired into the formalism, but ad hoc redefined from basic blocks of general automatic differentiation whenever it's needed.
+
+Mature Haskell libraries with similar capabilities, but varying efficiency, are https://hackage.haskell.org/package/ad and https://hackage.haskell.org/package/backprop. See also https://github.com/Mikolaj/horde-ad/blob/master/CREDITS.md.  
+The benchmarks at TODO show that this library has performance highly competitive with (i.e. faster than) those and PyTorch on CPU.  
+It is hoped that the separation of AD logic from matrix and tensor manipulation (deferred to [hmatrix] and [orthotope]) will enable similar speedups on numerical accelerators.
+
+## Computing the derivative of a simple function
+
+Here is an example of computing the gradient of a function that goes from `R^3` to `R`:
+
+```Haskell
+-- A function that goes from `R^3` to `R`. 
+-- [Arguments are tupled, could just as easily be curried. -awf]
+foo :: RealFloat a => (a,a,a) -> a
+foo (x,y,z) = 
+  let w = x * sin y 
+  in atan2 z w + z * w  -- [note extra w, we *want* to require returnLet up front --awf] 
+```
+
+In order to compute the gradient, we transform the function in the classic manner for forward-mode AD, to take `DualNumber` variables, 
+and the `DualMonad` which implements the system in the paper
+```Haskell
+fooM :: DualMonad d r m 
+     => DualNumberVariables d r -> m (DualNumber d r) 
+fooM variables = do { 
+     let x : y : z : _ = vars variables 
+     w <- x * sin y 
+     returnLet $ (atan2 z w + z * w)  -- [note extra w, we *want* to require returnLet up front --awf] 
+   }
+```
+We hope to automate this transfomation soon, automatically generating `fooM` from `foo`, but for now it's a mechanical, if somewhat tedious, manual task.
+
+And then the gradient of `foo` is
+```Haskell
+grad_foo :: HasDelta r => Domain0 r -> (Domain0 r, r) 
+grad_foo ds = 
+   let ((result, _, _, _), value) = 
+         dReverse 1 fooM (ds, V.empty, V.empty, V.empty) 
+   in (result, value) 
+```
+
+As can be verified by computing the gradient at `(1.1,2.2,3.3)`:
+```Haskell
+>>> grad_foo (mkDual [1.1, 2.2, 3.3])
+[7.75, 1.23, 5.52]
```

Comment:
This should remain a 2-liner, with as little line noise as possible. 
E.g. `grad_foo (DualVector.fromList [1.1 :: Float, 2.2, 3.3]`)